include/LinearSolver.h

```diff
@@ -54,6 +55,79 @@ namespace nalu{
 class LinearSolvers;
 class Simulation;
 
+/** LocalGraphArrays is a helper class for building the arrays describing
+ * the local csr graph, rowPointers and colIndices. These arrays are passed
+ * to the TpetraCrsGraph::setAllIndices method. This helper class is used
+ * within nalu's TpetraLinearSystem class.
+ * See unit-tests in UnitTestLocalGraphArrays.C.
+ */
+class LocalGraphArrays {
+public:
+  static const LocalOrdinal INVALID = std::numeric_limits<LocalOrdinal>::max();
+
+  LocalGraphArrays(const Kokkos::View<size_t*,sierra::nalu::DeviceSpace>& rowLengths)
+  : rowPointers(Teuchos::arcp<size_t>(rowLengths.size()+1)),
+    colIndices()
+  {
+    size_t nnz = compute_row_pointers(rowPointers, rowLengths);
+    colIndices = Teuchos::arcp<LocalOrdinal>(nnz);
+    for(int i=0, iend=colIndices.size(); i<iend; ++i) { colIndices[i] = INVALID; }
+  }
+
+  size_t get_row_length(size_t localRow) const { return rowPointers[localRow+1]-rowPointers[localRow]; }
```

Comment:
You're assuming UVM here.

Reply:
Be careful w/ "HostSpace" -- if that's a memory space, its default execution space might not be the same as the default execution space you intend to use.

You could leave these as device Views -- Tpetra has a "getEntryOnHost" function that you could copy for this purpose.